From ba76562094f5f5d4aac9390e5a2b4b6981cc608a Mon Sep 17 00:00:00 2001
From: Eric Yang <eric.yang.wiwynn@gmail.com>
Date: Thu, 2 Oct 2025 08:50:56 +0000
Subject: [PATCH 18/20] requester: Always re-initialize MCTP info
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

When receiving MCTP EID Interface Added and Connection Available,
re-initialize the MCTP info regardless of whether it has already been
initialized.

... after "Handle InterfaceAdded" ...
+----------------------------+
| EID in existingMctpInfos   |
+-------------+--------------+
              |
+-------------v--------------+
| remove MCTP info           |
+-------------+--------------+
              |
+-------------v--------------+
| re-init terminus           |
+----------------------------+

Motivation:
Currently, when pldmd receives an MCTP EID Interface Removed signal, it
calls GetSubtree to obtain the current set of MCTP EIDs, then compares
the result with existingMctpInfos to identify which MCTP info should
stop being managed, avoiding maintenance of a nonfunctional terminus.

However, under heavy BMC load, an EID may have been removed and even
re-added before pldmd processes the Interface Removed signal. As a
result, GetSubtree may show nothing to remove. Later, when handling
Interface Added, even though an EID shouldn’t be added unless it was
removed, pldmd skips terminus initialization, leaving a terminus in a
bad state.

[High-load scenario (BMC busy)]

+----------------------------+
| EID removed                |
| -> InterfaceRemoved (late) |
+-------------+--------------+
              |
+-------------v--------------+
| EID added                  |
| -> InterfaceAdded (queued) |
+-------------+--------------+
              |
+-------------v--------------+
| Handle InterfaceRemoved    |
+-------------+--------------+
              |
+-------------v--------------+
| GetSubtree: EID present    |
+-------------+--------------+
              |
+-------------v--------------+
| No EID to remove           |
+-------------+--------------+
              |
+-------------v--------------+
| Handle InterfaceAdded      |
+-------------+--------------+
              |
+-------------v--------------+
| EID in existingMctpInfos   |
+-------------+--------------+
              |
+-------------v--------------+
| skip init (BUG: stale)     |
+----------------------------+

Therefore, during MCTP endpoint discovery, if the MCTP info being
processed for Interface Added is already present in existingMctpInfos,
first run the remove flow for that MCTP info to ensure it is
re-initialized.

Change-Id: Ieaed380ed9dfa5f059bb04daf3e6105f76b1a0a9
Signed-off-by: Eric Yang <eric.yang.wiwynn@gmail.com>
---
 requester/mctp_endpoint_discovery.cpp | 46 +++++++++++++++++++++++++++
 requester/mctp_endpoint_discovery.hpp |  6 ++++
 2 files changed, 52 insertions(+)

diff --git a/requester/mctp_endpoint_discovery.cpp b/requester/mctp_endpoint_discovery.cpp
index 711c5da..e8bfdf9 100644
--- a/requester/mctp_endpoint_discovery.cpp
+++ b/requester/mctp_endpoint_discovery.cpp
@@ -291,6 +291,50 @@ void MctpDiscovery::removeFromExistingMctpInfos(MctpInfos& mctpInfos,
     }
 }
 
+void MctpDiscovery::removeBeforeAdd(const MctpInfos& toAddMctpInfo)
+{
+    if (toAddMctpInfo.empty())
+    {
+        return;
+    }
+
+    MctpInfos removedInfos;
+    for (const auto& existingInfo : existingMctpInfos)
+    {
+        const auto existingEid = std::get<0>(existingInfo);
+        const auto existingNet = std::get<3>(existingInfo);
+
+        auto it =
+            std::find_if(toAddMctpInfo.begin(), toAddMctpInfo.end(),
+                         [existingEid, existingNet](const auto& mctpInfo) {
+                             return std::get<0>(mctpInfo) == existingEid &&
+                                    std::get<3>(mctpInfo) == existingNet;
+                         });
+        if (it != toAddMctpInfo.end())
+        {
+            removedInfos.emplace_back(existingInfo);
+        }
+    }
+
+    if (removedInfos.empty())
+    {
+        return;
+    }
+
+    for (const auto& mctpInfo : removedInfos)
+    {
+        info(
+            "Removing Endpoint networkId '{NETWORK}' and EID '{EID}' to prevent duplicate addition of the same MCTP Info",
+            "NETWORK", std::get<3>(mctpInfo), "EID", std::get<0>(mctpInfo));
+        existingMctpInfos.erase(std::remove(existingMctpInfos.begin(),
+                                            existingMctpInfos.end(), mctpInfo),
+                                existingMctpInfos.end());
+    }
+
+    handleRemovedMctpEndpoints(removedInfos);
+    removeConfigs(removedInfos);
+}
+
 void MctpDiscovery::propertiesChangedCb(sdbusplus::message_t& msg)
 {
     using Interface = std::string;
@@ -342,6 +386,7 @@ void MctpDiscovery::propertiesChangedCb(sdbusplus::message_t& msg)
             {
                 if (availability)
                 {
+                    removeBeforeAdd(MctpInfos(1, mctpInfo));
                     // The endpoint not in existingMctpInfos and is
                     // available Add it to existingMctpInfos
                     info(
@@ -365,6 +410,7 @@ void MctpDiscovery::discoverEndpoints(sdbusplus::message_t& msg)
 {
     MctpInfos addedInfos;
     getAddedMctpInfos(msg, addedInfos);
+    removeBeforeAdd(addedInfos);
     addToExistingMctpInfos(addedInfos);
     handleMctpEndpoints(addedInfos);
 }
diff --git a/requester/mctp_endpoint_discovery.hpp b/requester/mctp_endpoint_discovery.hpp
index 7692849..e48a339 100644
--- a/requester/mctp_endpoint_discovery.hpp
+++ b/requester/mctp_endpoint_discovery.hpp
@@ -166,6 +166,12 @@ class MctpDiscovery
     void removeFromExistingMctpInfos(MctpInfos& mctpInfos,
                                      MctpInfos& removedInfos);
 
+    /** @brief Remove any existing MCTP endpoint with same Network
+     *         Index and Endpoint ID before adding the new MCTP endpoint.
+     *  @param[in] toAddMctpInfo - information of the new MCTP endpoint
+     */
+    void removeBeforeAdd(const MctpInfos& toAddMctpInfo);
+
     friend class ::TestMctpDiscovery;
 
   private:
-- 
2.43.0

